apps/passport-server/src/services/poapService.ts

```diff
@@ -0,0 +1,253 @@
+import { getEdDSAPublicKey } from "@pcd/eddsa-pcd";
+import { getHash } from "@pcd/passport-crypto";
+import { SerializedPCD } from "@pcd/pcd-types";
+import { ZKEdDSAEventTicketPCDPackage } from "@pcd/zk-eddsa-event-ticket-pcd";
+import AsyncLock from "async-lock";
+import { fetchDevconnectPretixTicketByTicketId } from "../database/queries/devconnect_pretix_tickets/fetchDevconnectPretixTicket";
+import {
+  claimNewPoapUrl,
+  getExistingClaimUrlByTicketId
+} from "../database/queries/poap";
+import { ApplicationContext } from "../types";
+import { logger } from "../util/logger";
+import { getServerErrorUrl } from "../util/util";
+import { RollbarService } from "./rollbarService";
+import { traced } from "./telemetryService";
+
+// Set up an async-lock to prevent concurrency issues when two threads
+// are simultaneously calling `getDevconnectPoapClaimUrl()`.
+const lock = new AsyncLock();
+const POAP_CLAIM_LOCK_KEY = "poap-claim-lock-key";
```

Comment:
Rather than having a global lock, I think you should have a per-ticket lock. So pass in the ticket id as the lock key parameter into the locking function. This way many users can concurrently ask for a POAP, but a single user can only ask for one at a time.